deltacat/compute/compactor/model/materialize_result.py

```diff
@@ -13,15 +14,19 @@ def of(
         delta: Delta,
         task_index: int,
         pyarrow_write_result: PyArrowWriteResult,
-        count_of_src_dfl_not_touched: Optional[int] = 0,
-        count_of_src_dfl: Optional[int] = 0,
+        referenced_pyarrow_write_result: Optional[PyArrowWriteResult] = None,
+        peak_memory_usage_bytes: Optional[np.double] = None,
+        telemetry_time_in_seconds: Optional[np.double] = None,
+        task_completed_at: Optional[np.double] = None,
```

Comment:
Since these 3 members are now common to all results, lets push them down into a `StepResult` base class that all materialize/hash-bucket/dedupe results extend.

Reply:
I wanted to do that but I was skeptical about the performance deterioration as we had observed that when we started returning dataclass. I believe composition is better than inheritance for reusing variables. Hence, clubbed them into a single object.

Reply:
The job now is taking 10 mins more to complete. I believe it's due to additional overhead due to nested tuple. Hence, I will keep it as is.